docs: document ESLint migration codemods in v9 and v10 guides (#20980)
eslint/eslint: docs/src/use/migrate-to-9.0.0.md

title: Migrate to v9.x
---

{%- from 'components/npx_tabs.macro.html' import npx_tabs %}

> 

Reviewer: ```suggestion
```

Same as https://github.com/eslint/eslint/pull/20980#discussion_r3408231494, I think this line is unused.

## Use migration codemods

### Migrate ESLint configuration

The `@eslint/v8-to-v9-config` codemod migrates ESLint v8 configuration files to the ESLint v9 flat config format. It updates config structure, rule schemas, plugins, ignores, and deprecated JSDoc rules while preserving existing behavior as much as possible.

```shell
npx codemod @eslint/v8-to-v9-config
```

Learn more in the [Codemod Registry](https://app.codemod.com/registry/@eslint/v8-to-v9-config).

### Migrate custom rules

If you maintain custom ESLint rules, the `@eslint/v8-to-v9-custom-rules` codemod converts function-style rules to the object format and updates deprecated rule APIs to their ESLint v9-compatible equivalents.

```shell
npx codemod @eslint/v8-to-v9-custom-rules
```

Learn more in the [Codemod Registry](https://app.codemod.com/registry/@eslint/v8-to-v9-custom-rules).

Codemods are a starting point. Review the changes and consult the breaking changes below for anything the codemods do not cover.

> The lists below are ordered roughly by the number of users each change is expected to affect, where the first items are expected to affect the most users.

Reviewer: ```suggestion
The lists below are ordered roughly by the number of users each change is expected to affect, where the first items are expected to affect the most users.

## Table of Contents
```

I think the `## Table of Contents` was removed by mistake here?

title: Migrate to v9.x
---

{%- from 'components/npx_tabs.macro.html' import npx_tabs %}

ESLint v9.0.0 is a major release of ESLint, and as such, has several breaking changes that you need to be aware of. This guide is intended to walk you through the breaking changes.

> To help with this migration, ESLint provides codemods to automate many of the changes described in this guide. All official ESLint codemods are available in the [eslint/codemods](https://github.com/eslint/codemods) repository and through the [Codemod Registry](https://codemod.link/eslint).

Reviewer: ```suggestion
To help with this migration, ESLint provides codemods to automate many of the changes described in this guide. All official ESLint codemods are available in the [eslint/codemods](https://github.com/eslint/codemods) repository and through the [Codemod Registry](https://app.codemod.com/registry?q=scope%3Aeslint).
```

## Use migration codemods

### Migrate ESLint configuration

The `@eslint/v8-to-v9-config` codemod migrates ESLint v8 configuration files to the ESLint v9 flat config format. It updates config structure, rule schemas, plugins, ignores, and deprecated JSDoc rules while preserving existing behavior as much as possible.

```shell
npx codemod @eslint/v8-to-v9-config
```

Learn more in the [Codemod Registry](https://app.codemod.com/registry/@eslint/v8-to-v9-config).

### Migrate custom rules

If you maintain custom ESLint rules, the `@eslint/v8-to-v9-custom-rules` codemod converts function-style rules to the object format and updates deprecated rule APIs to their ESLint v9-compatible equivalents.

```shell
npx codemod @eslint/v8-to-v9-custom-rules
```

Learn more in the [Codemod Registry](https://app.codemod.com/registry/@eslint/v8-to-v9-custom-rules).

Codemods are a starting point. Review the changes and consult the breaking changes below for anything the codemods do not cover.

The lists below are ordered roughly by the number of users each change is expected to affect, where the first items are expected to affect the most users.

## Table of Contents

> - [Use migration codemods](#use-migration-codemods)

Reviewer: Same here. I think removing it, or moving `## Use migration codemods` after the TOC, would be nice.

Author: removed TOC here.

ESLint v9.0.0 is a major release of ESLint, and as such, has several breaking changes that you need to be aware of. This guide is intended to walk you through the breaking changes.

To help with this migration, ESLint provides codemods to automate many of the changes described in this guide. All official ESLint codemods are available in the [eslint/codemods](https://github.com/eslint/codemods) repository and through the [Codemod Registry](https://app.codemod.com/registry?q=scope%3Aeslint).

## Use migration codemods

### Migrate ESLint configuration

> The `@eslint/v8-to-v9-config` codemod migrates ESLint v8 configuration files to the ESLint v9 flat config format. It updates config structure, rule schemas, plugins, ignores, and deprecated JSDoc rules while preserving existing behavior as much as possible.

Reviewer 2: The description of the config codemod is incomplete. In addition to config structure/rules/plugins/ignores/JSDoc migration, it also modifies package.json entries containing eslintConfig, .eslintignore/.gitignore files, /* exported */ comments, and certain /* eslint */ comments.

```shell
npx codemod @eslint/v8-to-v9-config
```

Learn more in the [Codemod Registry](https://app.codemod.com/registry/@eslint/v8-to-v9-config).

### Migrate custom rules

> If you maintain custom ESLint rules, the `@eslint/v8-to-v9-custom-rules` codemod converts function-style rules to the object format and updates deprecated rule APIs to their ESLint v9-compatible equivalents.

Reviewer 2: A caveat should be added for custom rules. The workflow runs on **/*.js/mjs/cjs/jsx/ts/mts/cts/tsx, and the README warns that it should only be executed against rule files/directories, since it may incorrectly transform non-rule JavaScript files.

- Node.js v20.9.0 and above
- Node.js v21 and above

> **Codemod:** The migration codemods do not cover this change.

Reviewer 2: No need to mention this, it's not related to both config file migration and custom rules migration

As announced in our [blog post](/blog/2023/10/flat-config-rollout-plans/), in ESLint v9.0.0, [`eslint.config.js`](./configure/configuration-files) is the new default configuration format. The previous format, eslintrc, is now deprecated and will not automatically be searched for.

**Codemod:** Use the [@eslint/v8-to-v9-config](#use-migration-codemods) codemod to automate much of this migration.

> **To address:** Update your configuration manually following the [Configuration Migration Guide](./configure/migration-guide). In case you still need to use the deprecated eslintrc config format, set environment variable `ESLINT_USE_FLAT_CONFIG` to `false`.

Reviewer: ```suggestion
**To address:** Update your configuration manually to the new format following the [Configuration Migration Guide](./configure/migration-guide). In case you still need to use the deprecated eslintrc config format, set environment variable `ESLINT_USE_FLAT_CONFIG` to `false`.
```

To preserve the original description.

| `unix`                | `eslint-formatter-unix`         |
| `visualstudio`        | `eslint-formatter-visualstudio` |

> **Codemod:** The migration codemods do not cover this change.

Reviewer 2: Fully supported

The `require-jsdoc` and `valid-jsdoc` rules have been removed in ESLint v9.0.0. These rules were initially deprecated in 2018.

> **Codemod:** The [@eslint/v8-to-v9-config](#use-migration-codemods) codemod removes deprecated JSDoc rules from your configuration.

Reviewer 2: “removes deprecated JSDoc rules” is not precise. The config codemod does more than remove them: when those rules are active in the configuration, it removes require-jsdoc/valid-jsdoc, adds eslint-plugin-jsdoc and jsdoc({ config: 'flat/recommended' }), and also removes file-level JSDoc ESLint comments.

- [`no-mixed-spaces-and-tabs`](../rules/no-mixed-spaces-and-tabs)
- [`no-new-symbol`](../rules/no-new-symbol)

> **Codemod:** The migration codemods do not cover this change.

Reviewer 2: ```suggestion
**Codemod:** The [@eslint/v8-to-v9-config](#use-migration-codemods) codemod partially covers this change. When your configuration extends `eslint:recommended`, it disables the four newly enabled recommended rules listed above (unless you already configure them explicitly) to preserve ESLint v8 lint behavior. It does not re-enable rules removed from the preset, fix new violations in source files, or change configs that use `@eslint/js` presets without a legacy `eslint:recommended` extend entry.

```

If `--max-warnings` is used then `--quiet` will not suppress the execution of rules set to `"warn"` but the output of those rules will be suppressed.

> **Codemod:** The migration codemods do not cover this change.

Reviewer 2: no need to mention, this is a CLI flag not related to codemod

Prior to ESLint v9.0.0, the `--output-file` flag would skip writing a file to disk if the output was empty. However, in ESLint v9.0.0, `--output-file` now consistently writes a file to disk, even when the output is empty. This update ensures a more consistent and reliable behavior for `--output-file`.

> **Codemod:** The migration codemods do not cover this change.

Reviewer 2: no need to mention, this is a CLI flag not related to codemod

- **Flat config.** If you are using flat config, you can run `npx eslint` or `eslint` (if globally installed) and ESLint will assume you want to lint the current directory. Effectively, passing no patterns is equivalent to passing `.`.
- **eslintrc.** If you are using the deprecated eslintrc config, you'll now receive an error when running the CLI without any patterns.

> **Codemod:** The migration codemods do not cover this change.

Reviewer 2: no need to mention, this is a CLI functionality change not related to codemod

Note that this change only affects cases where the same rule is configured in the config file with options and using a configuration comment without options. In all other cases (e.g. the rule is only configured using a configuration comment), the behavior remains the same as prior to ESLint v9.0.0.

> **Codemod:** The migration codemods do not cover this change.

Reviewer 2: no need to mention, because this is just a new runtime method in ESLint not related to codemod

foo(); // error: Missing semicolon
```

> **Codemod:** The migration codemods do not cover this change.

Reviewer 2: “codemods do not cover” duplicate `/* eslint */` comments is incorrect. The config workflow includes a remove-unnecessary-eslint-comments step that modifies these comments. Its actual behavior is broader than described: it removes all ESLint config comments matching its regex except the first one, not just duplicates of the same rule.

In ESLint v9.0.0, any `/* exported */` variables followed by a colon and value will be ignored as invalid.

> **Codemod:** The migration codemods do not cover this change.

Reviewer 2: “codemods do not cover” `/* exported */` comments is incorrect. The config codemod explicitly rewrites both `/* exported foo: true, bar: false */` and the whitespace-separated form into a comma-separated format.

foo - 0;
```

> **Codemod:** The migration codemods do not cover this change.

Reviewer 2: No need to mention, because this is just a new runtime method in ESLint not related to codemod

In ESLint v9.0.0, the option `allowConstructorFlags` is now case-sensitive.

> **Codemod:** The [@eslint/v8-to-v9-config](#use-migration-codemods) codemod may update your configuration for this.

Reviewer 2: ```suggestion
**Codemod:** The [@eslint/v8-to-v9-config](#use-migration-codemods) codemod expands `allowConstructorFlags` with opposite-case variants to preserve ESLint v8 behavior under v9's case-sensitive matching.

```

In ESLint v9.0.0, the option `allowConstructorFlags` is now case-sensitive.

**Codemod:** The [@eslint/v8-to-v9-config](#use-migration-codemods) codemod may update your configuration for this.

> **To address:** Update your configuration manually if needed.

Reviewer: ```suggestion
**To address:** Update your configuration manually if needed.

**Related issue(s):** [#16574](https://github.com/eslint/eslint/issues/16574)
```

I think the original related issue section was removed by mistake.

}
```

> **Codemod:** The migration codemods do not cover this change.

Reviewer 2: ```suggestion
**Codemod**: The config codemod copies varsIgnorePattern to caughtErrorsIgnorePattern when the rule is explicitly configured, catch errors are checked (caughtErrors is not "none"), and no separate caughtErrorsIgnorePattern is set. This preserves v8 behavior where varsIgnorePattern incorrectly applied to catch bindings. It does not apply when the rule comes only from a preset, when caughtErrors is "none", or for @typescript-eslint/no-unused-vars.
```

In previous versions of ESLint, with this configuration only `import { View } from "react-native"` would be reported.

> **Codemod:** The migration codemods do not cover this change.

Reviewer 2: ```suggestion
**Codemod:** The [@eslint/v8-to-v9-config](#use-migration-codemods) codemod updates `no-restricted-imports` configurations during migration. When the rule uses a `paths` array, it restructures the options for flat config and deduplicates entries that share the same `name`, keeping the last entry so lint behavior matches ESLint v8 (where only the last matching entry applied).

```

In ESLint v9.0.0, this format is no longer supported and will result in an error.

> **Codemod:** Use the [@eslint/v8-to-v9-config](#use-migration-codemods) codemod to migrate these strings to the `@eslint/js` package.

Reviewer 2: ```suggestion
**Codemod:** Use the [@eslint/v8-to-v9-config](#use-migration-codemods) codemod. It converts legacy string presets in existing `eslint.config.*` files and migrates eslintrc configs via `FlatCompat` + `@eslint/js`.

```

}
```

> **Codemod:** The migration codemods do not cover this change.

Reviewer 2: ```suggestion
**Codemod**: The config codemod adds blockScopedFunctions: "disallow" when no-inner-declarations is explicitly configured, preserving v8 reporting behavior. It does not apply when the rule comes only from a preset.
```

}
```

> **Codemod:** The migration codemods do not cover this change.

Reviewer 2: ```suggestion
**Codemod:** The [@eslint/v8-to-v9-config](#use-migration-codemods) codemod updates `no-unused-vars` when the rule is explicitly configured. It adds `caughtErrors: "none"` if that option is omitted, preserving ESLint v8 behavior. Existing `caughtErrors` values are kept unchanged. If catch errors are checked and `varsIgnorePattern` is set, it also copies that pattern to `caughtErrorsIgnorePattern`. It does not add the rule when it is enabled only through `eslint:recommended`, and it does not migrate `@typescript-eslint/no-unused-vars`.

```

}
```

> **Codemod:** The migration codemods do not cover this change.

Reviewer 2: This covered by codemod, if no-useless-computed-key was empty it will put `false` for default value

Previously the camelcase rule didn't enforce the `allow` option to be an array of strings. In ESLint v9.0.0, the `allow` option now only accepts an array of strings.

**Codemod:** The [@eslint/v8-to-v9-config](#use-migration-codemods) codemod may update your configuration for this.

> **To address:** If ESLint still reports invalid configuration for this rule, update your configuration manually.

Reviewer: ```suggestion
**To address:** If ESLint still reports invalid configuration for this rule, update your configuration manually.

**Related issue(s):** [#18232](https://github.com/eslint/eslint/pull/18232)
```

Same here.

| `context.getScope()`               | `sourceCode.getScope(node)`                 |
| `context.markVariableAsUsed(name)` | `sourceCode.markVariableAsUsed(name, node)` |

**Codemod:** Use the [@eslint/v8-to-v9-custom-rules](#use-migration-codemods) codemod to automate much of this migration.

**To address:** For more information, see the [blog post](https://eslint.org/blog/2023/09/preparing-custom-rules-eslint-v9/#automatically-update-your-rules).

**Related Issues(s):** [#16999](https://github.com/eslint/eslint/issues/16999), [#13481](https://github.com/eslint/eslint/issues/13481)

## <a name="removed-sourcecode-getcomments"></a> Removed `sourceCode.getComments()`

ESLint v9.0.0 removes the deprecated `sourceCode.getComments()` method.

> **Codemod:** Use the [@eslint/v8-to-v9-custom-rules](#use-migration-codemods) codemod to automate this change.

Reviewer 2: ```suggestion
**Codemod:** Use the [@eslint/v8-to-v9-custom-rules](#use-migration-codemods) codemod. It converts both `context.getComments()` and `sourceCode.getComments()` to combinations of `getCommentsBefore()`, `getCommentsInside()`, and `getCommentsAfter()`.

```

ESLint v9.0.0 now precalculates code path information before the traversal used by rules. As a result, the code path information is now complete regardless of where it is accessed inside of a rule.

> **Codemod:** The migration codemods do not cover this change.

Reviewer 2: ```suggestion
**Codemod**: The migration codemods do not cover code path precalculation. Rules that rely on when CodePath or CodePathSegment array properties are populated must be reviewed manually. The @eslint/v8-to-v9-custom-rules codemod only migrates removed CodePath#currentSegments usage.
```

As of ESLint v9.0.0, an error will be thrown if any options are [passed](../use/configure/rules#use-configuration-files) to a rule that doesn't specify `meta.schema` property.

> **Codemod:** The migration codemods do not cover this change.

Reviewer 2: “meta.schema is not covered” is not entirely accurate. The custom-rules codemod generates meta.schema for function-style and object-style rules, migrates legacy module.exports.schema, and inserts a TODO when it detects context.options without a schema. The limitation is that it does not fully infer the actual schema.

As announced in our [blog post](/blog/2023/10/flat-config-rollout-plans/), the temporary `FlatRuleTester` class has been renamed to `RuleTester`, while the `RuleTester` class from v8.x has been removed. Additionally, the `FlatRuleTester` export from `eslint/use-at-your-own-risk` has been removed.

> **Codemod:** The migration codemods do not cover this change.

Reviewer 2: ```suggestion
**Codemod**: The @eslint/v8-to-v9-custom-rules codemod renames FlatRuleTester to RuleTester and moves parserOptions to languageOptions in test cases. Updating imports from eslint/use-at-your-own-risk to eslint, adjusting for new RuleTester defaults, and translating other eslintrc-style test config must be done manually.
```

1. **`filename` and `only` must be of the expected type.** `RuleTester` now checks the type of `filename` and `only` properties of test objects. If specified, `filename` must be a string value. If specified, `only` must be a boolean value.
1. **Messages cannot have unsubstituted placeholders.** The `RuleTester` now also checks if there are {% raw %}`{{ placeholder }}` {% endraw %} still in the message as their values were not passed via `data` in the respective `context.report()` call.

> **Codemod:** The migration codemods do not cover this change.

Reviewer 2: ```suggestion
**Codemod**: The @eslint/v8-to-v9-custom-rules codemod renames FlatRuleTester to RuleTester, moves parserOptions to languageOptions in test cases, and removes output when it equals code. All other stricter RuleTester checks must be fixed manually by running your rule tests and updating failing cases.
```

As announced in our [blog post](/blog/2023/10/flat-config-rollout-plans/), the temporary `FlatESLint` class has been renamed to `ESLint`, while the `ESLint` class from v8.x has been renamed to `LegacyESLint`.

> **Codemod:** The migration codemods do not cover this change.

Reviewer 2: It's supported and `FlatESLint` will moved to `ESLint`

Additionally, methods `Linter#defineRule()`, `Linter#defineRules()`, `Linter#defineParser()`, and `Linter#getRules()` are no longer available.

> **Codemod:** The migration codemods do not cover this change.

Reviewer 2: ```suggestion
**Codemod**: The @eslint/v8-to-v9-custom-rules codemod renames FlatESLint to ESLint and moves parserOptions to languageOptions in Linter#verify() and Linter#verifyAndFix() calls. Calls to removed methods (defineRule, defineRules, defineParser, getRules) are marked with TODO comments; migrating rules and parsers into flat plugins / languageOptions config must be done manually.
```
